MAINT: stats.mode: remove deprecated features, smooth edges

mdhaber opened the pull request:
#### Reference issue
Closes gh-9955
Closes gh-15756
Closes gh-17214

#### What does this implement/fix?
This executes the removal of deprecated features and fixes bugs in `keepdims` and `nan_policy` behaviors of `stats.mode`.

mdhaber commented on scipy/stats/_axis_nan_policy.py:
```diff
@@ -356,7 +356,18 @@ def _axis_nan_policy_factory(tuple_to_result, default_axis=0,
     """
+    # Specify which existing behaviors the decorator must override
+    temp = override or {}
+    override = {'vectorization': False,
+                'nan_propagation': True}
```
These have been the default behaviors for the decorator: 
1. Don't override vectorization if `axis` is already an argument of the function.
2. Do override behavior for `nan_policy='propagate'` by returning `np.nan` for all elements of outputs.

This isn't appropriate for `mode`. 
1. Even though `mode` had an `axis` argument before, it was done with an N-d loop just like the decorator, so we might as well simplify the implementation and let the decorator take care of vectorization.
2. `np.nan` is not always the correct output when `nan_policy='propagate'`. For example, the mode of `[np.nan, 1, 1]` is `1`.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -496,7 +496,21 @@ def pmean(a, p, *, axis=0, dtype=None, weights=None):
+        count = count.dtype(0) if i else count
+    else:
+        count[i] = 0
+    return ModeResult(mode, count * (~np.isnan(count)))
```
Oops, this can be simplified. 
```suggestion
    return ModeResult(mode, count)
```
I was playing around with removing all of the code above and just doing `count * (~np.isnan(count))`, but `np.nan * 0` is `np.nan`, not `0`.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -542,21 +545,14 @@ def mode(a, axis=0, nan_policy='propagate', keepdims=None):
-    .. deprecated:: 1.9.0
-        Support for non-numeric arrays has been deprecated as of SciPy 1.9.0
-        and will be removed in 1.11.0. `pandas.DataFrame.mode`_ can
-        be used instead.
```
I will improve the error message when the user passes a non-numeric array in an immediate follow-up PR because this one is already quite large.
```python3
stats.mode(['hi', 'hi', 'bye'])
```
results in:
```
Traceback (most recent call last):

  File "C:\Users\matth\AppData\Local\Temp\ipykernel_15996\345230451.py", line 1, in <module>
    stats.mode(['hi', 'hi', 'bye'])

  File "C:\Users\matth\scipy\scipy\stats\_axis_nan_policy.py", line 513, in axis_nan_policy_wrapper
    res = hypotest_fun_out(*samples, **kwds)

  File "C:\Users\matth\scipy\scipy\stats\_stats_py.py", line 583, in mode
    return ModeResult(modes[()], counts[()])

TypeError: string indices must be integers
```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -572,75 +568,19 @@ def mode(a, axis=0, nan_policy='propagate', keepdims=None):
 
     >>> stats.mode(a, axis=None, keepdims=True)
-    ModeResult(mode=[3], count=[5])
+    ModeResult(mode=[[3]], count=[[5]])
```
This is the problem report in gh-17214. I think I let it slip when I was adding `keepdims` because this had been the behavior of `mode` before `keepdims` was introduced.

mdhaber commented on scipy/stats/tests/test_axis_nan_policy.py:
```diff
@@ -138,7 +144,8 @@ def nan_policy_1d(hypotest, data1d, unpacker, *args, n_outputs=2,
                 raise ValueError("The input contains nan values")
 
-    elif nan_policy == 'propagate':
+    elif (nan_policy == 'propagate'
+          and hypotest not in override_propagate_funcs):
```
Note that `stats.mode not in override_propagate_funcs` is `False`, so none if the `if`/`elif` special cases apply, and the test relies on the behavior of `mode` for 1D slice to get the reference value (just like it would if there were no NaNs at all).

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2190,8 +2190,8 @@ def test_scalar(self):
         vals = stats.mode(data1)
-        assert_equal(vals[0][0], 6)
-        assert_equal(vals[1][0], 3)
+        assert_equal(vals[0], 6)
+        assert_equal(vals[1], 3)
```
The default behavior is no longer `keepdims`, so I'm removing a bunch of things like this.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2201,15 +2201,15 @@ def test_axes(self):
-        vals = stats.mode(arr, axis=None)
-        assert_equal(vals[0], np.array([30]))
-        assert_equal(vals[1], np.array([8]))
+        vals = stats.mode(arr, axis=None, keepdims=True)
+        assert_equal(vals[0], np.array([[30]]))
+        assert_equal(vals[1], np.array([[8]]))
```
In other cases, I'll add `keepdims=True`.
Here, I also had to change the test because of the inconsistency noted in gh-17214.

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -2412,21 +2345,66 @@ def test_gh16952(self):
+        ref = ([np.nan, np.nan], [0, 0])
+        assert_equal(res, ref)
+
+    @pytest.mark.filterwarnings('ignore::RuntimeWarning')  # np.mean warns
+    def test_gh9955b(self):
+        # A more thorough test of gh-9955 with empty arrays based on
+        # `test_axis_nan_policy.test_empty`
+
+        def small_sample_generator(n_dims):
+            # return all possible "small" arrays in up to n_dim dimensions
+            for i in n_dims:
+                # "small" means with size along dimension either 0 or 1
+                for combo in combinations_with_replacement([0, 1, 2], i):
+                    yield np.zeros(combo)
+
+        for sample in small_sample_generator([2, 3]):
+            # this test is only for arrays of zero size
+            if not sample.size == 0:
+                continue
+
+            # need to test for all valid values of `axis` parameter, too
+            for axis in range(-sample.ndim, sample.ndim):
+                res = stats.mode(sample, axis=axis)
+
+                ref = np.mean(sample, axis=axis) * np.nan  # mode is always NaN
+                assert_equal(res.mode, ref)
+                assert np.shape(res.mode) == ref.shape
+
+                ref[np.isnan(ref)] = 0  # count is always zero
+                assert_equal(res.count, ref)
+                assert res.count.shape == ref.shape
+
```
We don't necessarily need to include this test, but I wanted to make sure it passed.

mdhaber reviewed (commented):
Some self-review to aid the reviewer.

mdhaber commented:
@tirthasheshpatel just thought I'd check in - we finally get to fix this!

tirthasheshpatel commented on scipy/stats/tests/test_axis_nan_policy.py:
```diff
@@ -595,6 +602,10 @@ def _check_arrays_broadcastable(arrays, axis):
 
+    if hypotest in override_propagate_funcs:
+        reason = "Doesn't follow the usual pattern. Tested separately."
+        pytest.xfail(reason=reason)
```
I think we should skip instead of xfail.

tirthasheshpatel commented on scipy/stats/tests/test_axis_nan_policy.py:
```diff
@@ -73,6 +74,11 @@ def unpack_ttest_result(res):
+override_propagate_funcs = {stats.mode}
+
+# For some functions, empty input should not result in all-NaN output
+override_empty_funcs = {stats.mode}
```
This is unused. Is this like a note that `stats.mode` is different?

mdhaber replied:
Oops. I think that was from before a refactoring. Removed.

tirthasheshpatel commented on scipy/stats/_stats_py.py:
```diff
@@ -542,21 +545,14 @@ def mode(a, axis=0, nan_policy='propagate', keepdims=None):
     binary representations - are treated as equivalent and counted as separate
     instances of the same value.
 
+    By convention, the mode of an empty array is NaN, and the associated count
+    is zero.
```
> Is this a good convention?

It looks good to me.

tirthasheshpatel reviewed (commented):
Thanks @mdhaber! Sorry for the late review. The changes look pretty good! Just a small comment and I think we can merge. I also noted that there is a slight backward-incompatibility: previously, `nan_policy='omit'` returned masked arrays while now we just return normal arrays (see https://github.com/scipy/scipy/issues/9955#issuecomment-474166442). Was this discussed anywhere? Anyways, masked arrays were not useful here (since nothing was ever masked) and I am +1 for returning normal arrays.

mdhaber commented:
Thanks @tirthasheshpatel! I'll complete the rest when I'm at a computer. In the meantime, I think this addresses the comment about masked arrays:
https://docs.scipy.org/doc/scipy/dev/api-dev/nan_policy.html#how-not-to-implement-nan-policy

mdhaber commented:
Thanks @tirthasheshpatel. I made the adjustments you suggested. If it still looks good to you, it's ready from my perspective.

tirthasheshpatel reviewed (approved):
Thanks for the changes @mdhaber! Everything LGTM now. Not sure why Cirrus CI keeps failing. But I think it should be OK to merge.

j-bowhay commented:
@mdhaber would you be able to add a line about this under the expired deprecations section of the release notes?

mdhaber commented:
@j-bowhay done. I also adjusted several backticks according to Sphinx requirements:
- single backticks link to the documentation of a class, function, or method
- double backticks render in monospaced font, but do not link
